File: app/models/domain/product/Product.scala
package models.domain.product

import models.dao.ProductDAO
import models.domain.product.productImage.ProductImage
import models.domain.util.Date
import models.ebean.{Product => EProduct}
import play.api.libs.json.{Json, OFormat}

case class Product(id: Option[Long], name: String, imageUrl: String, description: String, uploadDate: Date, isValidated: Boolean, userId: Long, subcategoryId: Long) {
  def equals(product: Product): Boolean = {
    if(product.id.isDefined && id.isDefined) id.get.equals(product.id.get)
    else false
  }
}

object Product extends ProductJsonFormat {

  def apply(product: EProduct): Product = {
    Product(
      Option(product.getId),
      product.getName,
      product.getImageUrl,
      product.getDescription,
      Date.now,
      product.isValidated,
      product.getUserId,
      product.getSubcategoryId
    )
  }

  def apply(productCreate: ProductCreate): Product = {
    Product(
      None,
      productCreate.name,
      productCreate.imageUrl.get,
      productCreate.description,
      Date.now,
      productCreate.isValidated,
      productCreate.userId,
      productCreate.subcategoryId
    )
  }

  def saveOrUpdate(product: Product): Option[Product] = {
    ProductDAO.saveOrUpdate(product)
  }

  def getById(id : Long) : Option[Product] = {
    ProductDAO.getById(id)
  }

  def getByName(name : String, subcategoryId: Long) : Option[List[Product]] = {
    ProductDAO.getByName(name, subcategoryId)
  }

  def getAll: List[Product] = {
    ProductDAO.getAllProducts
  }

  def validate(product: Product): Product = {
      ProductDAO.validate(product)
  }

  def delete(product: Product): Option[Boolean] = {
    ProductDAO.delete(product)
  }

  def addImage(productImage: ProductImage): ProductImage = {
    ProductDAO.addImage(productImage)
  }

}

trait ProductJsonFormat{
  implicit val productFormat: OFormat[Product] = Json.format[Product]
}

File: app/models/ebean/Product.java
package models.ebean;

import io.ebean.Ebean;
import io.ebean.Finder;
import io.ebean.Model;
import io.ebean.annotation.NotNull;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.Id;
import java.util.List;
import java.util.Optional;
import org.joda.time.DateTime;

@Entity
public class Product extends Model {

    @Id
    @NotNull
    private Long id;
    @NotNull
    private String name;
    @NotNull
    private String imageUrl;
    @NotNull
    @Column(length = 900)
    private String description;
    private DateTime uploadDate;
    @NotNull
    private boolean isValidated;
    private Long userId;
    private Long subcategoryId;


    private static Finder<Long, Product> finder = new Finder<>(Product.class);

    public Product(Long id, String name, String imageUrl, String description, DateTime uploadDate, boolean isValidated, Long userId, Long subcategoryId) {
        this.id = id;
        this.name = name;
        this.imageUrl = imageUrl;
        this.description = description;
        this.uploadDate = uploadDate;
        this.isValidated = isValidated;
        this.userId = userId;
        this.subcategoryId = subcategoryId;
    }

    @Override
    public void save(){
        super.save();
    }

    public static Optional<Product> getProductById(Long id){
        Product product = finder.byId(id);
        if(product != null){
            return Optional.of(product);
        }
        return Optional.empty();
    }

    public static Optional<List<Product>> getProductBySubcategoryId(Long id){
        return Optional.of(Ebean.find(Product.class).where().eq("subcategory_id", id).findList());
    }

    public static Optional<List<Product>> getByName(String name, Long subcategoryId){
        return Optional.of(Ebean.find(Product.class).where().icontains("name", name).eq("subcategory_id", subcategoryId).eq("is_validated", 1).findList());
    }

    public static Integer getTotalProductAmountByUserId(Long id){
        return Ebean.find(Product.class).where().eq("user_id", id).findCount();
    }

    public static Integer getValidatedProductAmountByUserId(Long id){
        return Ebean.find(Product.class).where().eq("user_id", id).eq("is_validated", 1).findCount();
    }

    public static List<Product> getAllProducts() {
        return finder.all();
    }

    public Long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public String getImageUrl() {
        return imageUrl;
    }

    public String getDescription() {
        return description;
    }

    public DateTime getUploadDate() {
        return uploadDate;
    }

    public boolean isValidated() {
        return isValidated;
    }

    public Long getUserId() {
        return userId;
    }

    public Long getSubcategoryId() {
        return subcategoryId;
    }
}
